src/scripts/stores/componentManager.ts

```diff
@@ -0,0 +1,52 @@
+import { atom } from 'nanostores';
+
+export type ComponentElementInstance = {
+    uid: string;
+    name: string;
+    instance: ComponentElement;
+};
+
+export class ComponentElement extends HTMLElement {
+    public uid: string;
+
+    constructor() {
+        super();
+        this.uid = (() => {
+            $componentsManagerIncrement.set($componentsManagerIncrement.get() + 1);
+            return $componentsManagerIncrement.get().toString();
+        })();
+    }
+
+    // =============================================================================
+    // Lifecycle
+    // =============================================================================
+    connectedCallback() {
+        this.register();
+    }
+
+    disconnectedCallback() {
+        this.unregister();
+    }
+
+    // =============================================================================
+    // Methods
+    // =============================================================================
+    register() {
+        $componentsManager.set([
+            ...$componentsManager.get(),
+            {
+                uid: this.uid,
+                name: this.constructor.name,
+                instance: this
+            }
+        ]);
+    }
+    unregister() {
+        $componentsManager.set(
+            $componentsManager.get().filter((component) => component.uid !== this.uid)
+        );
+    }
+}
+
+export const $componentsManagerIncrement = atom<number>(0);
+export const $componentsManager = atom<ComponentElementInstance[]>([]);
```

Comment:
The following is in response to [my initial review](https://github.com/locomotivemtl/astro-boilerplate/pull/30#pullrequestreview-2383424462).

After refreshing my memory on `WeakMap`, `WeakRef`, and `WeakSet`, I realize I forgot quite a few important aspects of them. Notably the lack of iterability. I also reviewed and tested `disconnectedCallback()` and this works well in terms of garbage collection.

My suggestion below includes using `HTMLElement.id` as the unique ID and, more drasticly, storing the custom element directly in the store instead of relying on a wrapper object. I am unfamiliar, and have not tested, if nanostore handles DOM elements well.

```suggestion
export class ComponentElement extends HTMLElement {
    constructor() {
        super();

        if (this.id === '') {
            const index = $componentsManagerIncrement.get() + 1;
            $componentsManagerIncrement.set(index);
            this.id = `${this.constructor.name.toLowerCase()}-${index}`;
        }
    }

    // =============================================================================
    // Lifecycle
    // =============================================================================
    connectedCallback() {
        this.register();
    }

    disconnectedCallback() {
        this.unregister();
    }

    // =============================================================================
    // Methods
    // =============================================================================
    register() {
        $componentsManager.set([
            ...$componentsManager.get(),
            this
        ]);
    }
    unregister() {
        $componentsManager.set(
            $componentsManager.get().filter((component) => component.id !== this.id)
        );
    }
}

export const $componentsManagerIncrement = atom<number>(0);
export const $componentsManager = atom<HTMLElement[]>([]);
```

Using your examples in the manager's README:


```ts
import { $componentsManager } from '@root/src/scripts/stores/componentManager';

$componentsManager.get().forEach((component) => {
    if (component instanceof Foo) {
        const $foo = component;
        $foo.doSomething(); // Call method on each Foo instance
    }
});
```

```ts
const $foo = $componentsManager.get().find((component) => component.id === 'foo') as Foo;
if ($foo) {
    $foo.doSomething(); // Call method on the Foo instance
}
```